# Assigment 2 | Applied Machine Learning | Paras Ahuja

### 1. At a high-level, without entering into mathematical details, compare and contrast the following classifiers: 

* Perceptron: In the book Perceptron algorithm was used to minimize classification error. We are simply updating the weights until a class is correctly classified. It is rather simplistic in its approach. For Perceptron we randomly initialize weights, then we compute the linear combination of an example with randomly initialized weights, and finally this is passed to a threshold function, if our predicted class is correct, then we do not update the weights, however it is incorrect we compute the delta weight, which is $\eta \times(y^{(i)} - \hat{y}^{(i)}) \times x_j^i$, and simultaneously update the weights. Perceptron works well when classes are linearly separable. However, if they are not then we have to provide hyperparameters, which allow us to tune, and set a limit to the number of iterations before we stop. One of the hyperparameters is theta, which in the book has been randomly assigned. It is rather difficult to classify input vectors that are not linearly separable. Single layer Perceptron is very fast and computationally less expensive when it comes it simple problems, however for more complex problems other algorithms might be better suited.


* SVM: Support Vector Machine is considered as an extension of Perceptron. In SVM we want to maximize the margins around the hyperplane, which is the decision boundary between the the classes. The objective function or the optimization function for SVM is $\frac{2}{||w||}$, and our goal is to maximize this. This also means that we want to minimize $||w||$. However, in practice we minimize the term $\frac{1}{2}||w||^2$. SVM is robust in the sense that it helps us curtail some of the overfitting issues. When we maximize the margins, what we are really doing is ensuring that large margins tend to have lower generalization error, however models with small margins are more prone to overfitting the data. If we overfit the data we are in high variance state. SVM also works with non-linearly separable data. In this instance it projects the data onto a higher plane using kernel methods. SVM is definitely much more robust than Perceptron.


* Decision Tree: Decision tree classifiers have an advantage when it comes to interpretability of the data. First thing we optimize here is the information gain. We split the data on the node that results in the largest information gain. This process is repeated until we are only left with leaves. Also, note that there are 3 splitting criteria. We can use Gini impurity, entropy, and classification error. Decision trees are very robust, and we can build complex decision boundaries. Also there are libraries like graphviz available that allow us to visualize our tree. There is potential of overfitting as well as the tree gets deeper, so we can set maximum depth or start pruning the tree. 


* Random Forest: Random forsests stem from decision trees. These are also very robust models. The idea behind random forests is to avergae the decision trees that individually suffer from overfitting. Random forests therefore have a better generalization performance and are less susceptible to overfitting the data. In comparison to decision trees random forests don't offer the same level of interpretability. Furthermore, one of the biggest advantage of random forests is that we don't have to worry so much about choosing good hyperparameter values. We do need to pay a close attention to the number of trees we want in our random forest. 


Which one of these algorithms should be used first depends on the type of data we have, and what criteria matters. For instance, if we have information about the data that the data is not linearly separable, then we should not think of using Perceptron, and instead move forward with SVM, but if iterpretability matters, then we have to use decision tree classifier. Even then, if our data is prone to overfitting then we move on to random forests, keeping in mind that we will not have the same level of interpretability. Therefore, which algorithm should be used first is rather subjective. However, if we are simply given a dataset and asked to create a model, we should start with the simplest first and then move upwards, in this case we will start with the Perceptron.

### 2. Using real datasets (can also be hypothetically constructed by yourself) define the following feature types, and give example values from your dataset. How would you represent these features in a computer program? (e.g. 32-bit integer? Floating point? String?

* Numerical
* Nominal
* Date
* Text
* Image
* Dependent Variable

In [1]:
import pandas as pd
import numpy as np
import math
np.set_printoptions(suppress=True)

In [2]:
df = pd.read_csv("http://archive.ics.uci.edu/ml/machine-learning-databases/00267/data_banknote_authentication.txt")
attributes = ["variance of Wavelet Transformed image", "skewness of Wavelet Transformed image", "curtosis of Wavelet Transformed image", "entropy of image", "class"]
df.columns = attributes
print(f'N rows={len(df)}, M columns={len(df.columns)}')
df.head()

N rows=1371, M columns=5


variance of Wavelet Transformed image  \
0                                4.54590   
1                                3.86600   
2                                3.45660   
3                                0.32924   
4                                4.36840   

   skewness of Wavelet Transformed image  \
0                                 8.1674   
1                                -2.6383   
2                                 9.5228   
3                                -4.4552   
4                                 9.6718   

   curtosis of Wavelet Transformed image  entropy of image  class  
0                                -2.4586          -1.46210      0  
1                                 1.9242           0.10645      0  
2                                -4.0112          -3.59440      0  
3                                 4.5718          -0.98880      0  
4                                -3.9606          -3.16250      0

We notice that we have 525,461 instances in this data, and there are 8 features. Furthermore, there maybe some missing instances. UCI mostly marks the missing instances with a "?", and we will replace these and drop any rows with missing information. 

In [3]:
df = df.replace("?", np.nan)
df = df.dropna()
df.head()

variance of Wavelet Transformed image  \
0                                4.54590   
1                                3.86600   
2                                3.45660   
3                                0.32924   
4                                4.36840   

   skewness of Wavelet Transformed image  \
0                                 8.1674   
1                                -2.6383   
2                                 9.5228   
3                                -4.4552   
4                                 9.6718   

   curtosis of Wavelet Transformed image  entropy of image  class  
0                                -2.4586          -1.46210      0  
1                                 1.9242           0.10645      0  
2                                -4.0112          -3.59440      0  
3                                 4.5718          -0.98880      0  
4                                -3.9606          -3.16250      0

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1371 entries, 0 to 1370
Data columns (total 5 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   variance of Wavelet Transformed image  1371 non-null   float64
 1   skewness of Wavelet Transformed image  1371 non-null   float64
 2   curtosis of Wavelet Transformed image  1371 non-null   float64
 3   entropy of image                       1371 non-null   float64
 4   class                                  1371 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 64.3 KB


Notice that the data here is either nominal, ordinal, or binomial. We are not dealing with numeric data rather than a grading scale of sorts. Here is how these have been defined by UCI.

1. variance of Wavelet Transformed image (continuous)
2. skewness of Wavelet Transformed image (continuous)
3. curtosis of Wavelet Transformed image (continuous)
4. entropy of image (continuous)
5. class (integer) 


   * Numerical: Numerical variable can be represented as either an int type or a float. We can easily perform numerical calculations on this data type, and find meaningful information. In this dataset we have variance of Wavelet Transformed image, skewness of Wavelet Transformed image, and curtosis of Wavelet Transformed image as numerical variable. As we can see these are definitely in float64 format.
   * Nominal: Nominal data has to represented as an int type. Nominal data is defined as data that is used for naming or labelling variables, without any quantitative value. In our dataset, we do not have any nominal dataset.
   * Date: Pandas has a datetime64[ns] type, therefore it is most likely to represented as this type. 
   * Text: will be represented a string type. We can use text information to establish a certain criteria like string containing a certain word.
   * Image: We do not have image in our dataset, however it will represented as a matrix of number where the numbers would represent pixels. Similar to the one in module 1 digit analysis. 
   * Dependent Variable: Our dependent variable is generally a nominal data if we have a classification task, and it is numeric if the dataset is regression dataset.


### 3. Using online resources, research and find other classifier performance metrics which are also as common as the accuracy metric. Write down the mathematical equations and the meaning of the metrics that you found.

**Accuracy:**

This is the most commonly used metric, here we evaluate the total number of correct predictions (True Positive + True Negative) divided by the total number data examined (True Positive + False Positive + False Negative + True Negative). This is calculated as:

$\text{Accuracy} = \frac{TP+TN}{TP+FP+FN+TN}$

**Precision:**

Precision is a metric that we use when we are really interested in how much of our predictions are positive. We use precision metric when we want to be very certain about the prediction we have made. For instance, if we are dealing with a person's health information or financial data. We calculate this metric as: 

$\text{Precision} = \frac{TP}{TP+FP}$

**Recall:**

With recall we are in interested in find out what proportion of actual Positives is correctly classified. We can also use this in tandem with Precision. We calculate this metric as: 

$\text{Recall} = \frac{TP}{TP+FN}$

**F1 Score:**

F1 Score is a metric that finds a balance between precision and recall. We calculate this as: 

$F_1 = 2 \times \frac{Precision \times Recall}{Precision + Recall}$

F1 Score is also widely used in the field of information retrieval for measuring search, document classification, and query classification performance. 

**Log Loss:**

Log Loss measures the performance of a classification model where the prediction is a probability. As we know probability is always between 0, and 1, and multiple probabilities for the same event add up to 1. Log loss increases as the predicted probability diverge from the actual label. Log loss is calculated as: 

$\text{Log-loss} = \frac{-1}{N}\sum_{i=1}^{N}y_i\log{p_i} + (1-y_i)\log(1-p_i)$ 

**References:**

https://towardsdatascience.com/performance-metrics-for-classification-machine-learning-problems-97e7e774a007

https://towardsdatascience.com/metrics-to-evaluate-your-machine-learning-algorithm-f10ba6e38234

### 4. Implement a correlation program from scratch to look at the correlations between the features of Admission_Predict.csv dataset file (not provided, you have to download it by yourself by following the instructions in the module Jupyter notebook). Display the correlation matrix where each row and column are the features, which should be an 8 by 8 matrix (should we use 'Serial no'?). You can use pandas DataFrame.corr() to verify correctness of yours. Observe that the diagonal of this matrix should have all 1's and explain why? Since the last column can be used as the target (dependent) variable, what do you think about the correlations between all the variables? Which variable should be the most important for prediction of 'Chance of Admit'?

In [5]:
# Load the dataset
admission = pd.read_csv('../Module 2 Assignment/Admission_Predict.csv')

# Check the shape of the dataset
print(f'N rows={len(admission)}, M columns={len(admission.columns)}')

# Check the dataset
admission.head()

N rows=400, M columns=9


Serial No.  GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  \
0           1        337          118                  4  4.5   4.5  9.65   
1           2        324          107                  4  4.0   4.5  8.87   
2           3        316          104                  3  3.0   3.5  8.00   
3           4        322          110                  3  3.5   2.5  8.67   
4           5        314          103                  2  2.0   3.0  8.21   

   Research  Chance of Admit   
0         1              0.92  
1         1              0.76  
2         1              0.72  
3         1              0.80  
4         0              0.65

In [6]:
def force_numeric(df, exception):
    """ This function takes in a Pandas DataFrame as its parameter
        and traverses through the columns coerces each column value
        to numeric. Non-numeric values get converted to NaN.
        This method returns DataFrame. This method does not discriminate
        between nominal, ordinal, and numerical data. 
    """
    for column in df.columns:
        if column != exception:
            df[column] = pd.to_numeric(df[column], errors="coerce")
    return df

In [7]:
# Check of any missing values
admission = force_numeric(admission, "Chance of Admit")
admission.isnull().values.any()

False

Note that serial number is nominal data. It is simply listed in numbers, and it would not have made much difference if serial number was alphabetical or alphanumeric. Therefore, we can drop this column from our assessment. 

In [8]:
admission = admission.drop(columns="Serial No.")

In [9]:
admission_array = admission.values
print("First five rows:" + "\n" + "{}".format(admission_array[0:5]))

First five rows:
[[337.   118.     4.     4.5    4.5    9.65   1.     0.92]
 [324.   107.     4.     4.     4.5    8.87   1.     0.76]
 [316.   104.     3.     3.     3.5    8.     1.     0.72]
 [322.   110.     3.     3.5    2.5    8.67   1.     0.8 ]
 [314.   103.     2.     2.     3.     8.21   0.     0.65]]


Correlation formula is as follows:

$r_{xy} = \frac{\sum(x_i - \bar{x})(y_i - \bar{y})}{\sqrt{\sum(x_i - \bar{x})^2\sum(y_i - \bar{y})^2}}$

Where:

   * $r_{xy}$ represents the correlation coefficient of the linear relationship between x and y
   * $x_i$ represents the $i^{th}$ value of the x-variable in the sample
   * $\bar{x}$ represents the mean of the values of the x-variable
   * $y_i$ represents the $i^{th}$ value of the y-variable in the sample
   * $\bar{y}$ represents the mean of the values of the y-variable

In [10]:
def correlation(array):
    """ This function takes in array and calculates
        the correlation matrix. The function assumes
        that each row has the same number of columns. 
        Any missing values values are assumed to be 
        entred as 0. 
        
        Example: [[1, 2, 3], [4, 6]] is not a valid
        parameter. 
        
        Example: [[1, 2, 3], [4, 0, 6]] is a valid
        parameter.
    """
    array = np.array(array)
    
    # calculate covariance matrix                    
    covariance_matrix = np.cov(array.T)
        
    # calculate standard deviation
    standard_deviation = np.std(array, axis=0)
    
    standard_deviation_matrix = []
    
    k = 0
    while k < len(standard_deviation):
        std = standard_deviation[k]
        arr = []
        for i in range(len(standard_deviation)):
            multiply = std * standard_deviation[i]
            arr.append(multiply)
        standard_deviation_matrix.append(arr)
        k += 1
    
    standard_deviation_matrix = np.array(standard_deviation_matrix).T
    
    correlation_matrix = []
    
    for i in range(len(standard_deviation_matrix)):
        nil = []
        for j in range(len(standard_deviation_matrix)):
            nil.append(0)
        correlation_matrix.append(nil)
    
    for i in range(len(standard_deviation_matrix)):
        for j in range(len(standard_deviation_matrix)):
            correlation_matrix[i][j] = round(covariance_matrix[i][j] / standard_deviation_matrix[i][j], 2)
    
    return np.array(correlation_matrix).T

In [11]:
correlation = correlation(admission_array)

In [12]:
correlation

array([[1.  , 0.84, 0.67, 0.61, 0.56, 0.84, 0.58, 0.8 ],
       [0.84, 1.  , 0.7 , 0.66, 0.57, 0.83, 0.49, 0.79],
       [0.67, 0.7 , 1.  , 0.74, 0.66, 0.75, 0.45, 0.71],
       [0.61, 0.66, 0.74, 1.  , 0.73, 0.72, 0.45, 0.68],
       [0.56, 0.57, 0.66, 0.73, 1.  , 0.67, 0.4 , 0.67],
       [0.84, 0.83, 0.75, 0.72, 0.67, 1.  , 0.52, 0.88],
       [0.58, 0.49, 0.45, 0.45, 0.4 , 0.52, 1.  , 0.55],
       [0.8 , 0.79, 0.71, 0.68, 0.67, 0.88, 0.55, 1.  ]])

In [13]:
admission.corr()

GRE Score  TOEFL Score  University Rating       SOP  \
GRE Score           1.000000     0.835977           0.668976  0.612831   
TOEFL Score         0.835977     1.000000           0.695590  0.657981   
University Rating   0.668976     0.695590           1.000000  0.734523   
SOP                 0.612831     0.657981           0.734523  1.000000   
LOR                 0.557555     0.567721           0.660123  0.729593   
CGPA                0.833060     0.828417           0.746479  0.718144   
Research            0.580391     0.489858           0.447783  0.444029   
Chance of Admit     0.802610     0.791594           0.711250  0.675732   

                       LOR       CGPA  Research  Chance of Admit   
GRE Score          0.557555  0.833060  0.580391          0.802610  
TOEFL Score        0.567721  0.828417  0.489858          0.791594  
University Rating  0.660123  0.746479  0.447783          0.711250  
SOP                0.729593  0.718144  0.444029          0.675732  
LOR                1.000000  0.670211  0.396859          0.669889  
CGPA               0.670211  1.000000  0.521654          0.873289  
Research           0.396859  0.521654  1.000000          0.553202  
Chance of Admit    0.669889  0.873289  0.553202          1.000000

I have rounded the numbers of correlation matrix to two decimal places to present in a concise manner, however we can see that the diagonal of the matrix is all 1's, and other correlations also match with the ones provided by pandas library. If a student wants highest chance of admittance, then they need to focus on CGPA, and GRE Score to along with it. Notice that the Chance of Admit is 80% with GRE Score, and 87% with CGPA. However the variabe that should be most important is a student's CGPA. A correlation of 1 along the diagonal means perfect correlation, but notice that the diagonals are where a feature has a correlation with itself, and naturally a feature is perfectly correlated with itself.